### ISSUE TITLE
Enable to specify document uri

### ISSUE THREAD

**Issue by rmannibucau** [materiahq/ngx-monaco-editor#26]
Hi,

To be able to use document uri, it can be neat to be able to pass in options the uri of the document.
Typically it enables to have implicit schema validation and completion without having to match any document (*) or bet on "inmemory://model/1" implicit uri.

See https://github.com/microsoft/monaco-editor/issues/191 for details.

Romain

**Comment by jmarc-roy** [comment]
Hi @rmannibucau,

Thanks for this excellent suggestion.
I will investigate to see the best way to achieve this with the current implementation.

**Comment by calico-crusade** [comment]
Any update on this? A work around?
This is limiting my ability to use auto-complete as I cannot use external library type declarations without being able to set the URI.

**Comment by calico-crusade** [comment]
I found a work around for now, it isn't very pretty though:
In the component with your editor, you can do this:

```TypeScript
import { Component, OnInit } from '@angular/core';
import { MonacoEditorLoaderService } from '@materia-ui/ngx-monaco-editor';
//Using this specifically for getting typings, don't use it for actual code related things, as it isn't the same instance as window.monaco
import * as monacoEditor from 'monaco-editor';
import { filter, take } from 'rxjs/operators';

@Component({
	selector: 'cc-editor',
	templateUrl: './editor.component.html',
	styleUrls: ['./editor.component.scss']
})
export class EditorComponent implements OnInit {
	private editor: monacoEditor.editor.IStandaloneCodeEditor;
	editorOptions = {
		theme: 'vs-dark',
		language: 'typescript',
		minimap: {
			enabled: false
		},
		model: null
	};
	code: string = `...`;

	constructor(
		private _monacoSer: MonacoEditorLoaderService
	) {
		this._monacoSer.isMonacoLoaded$.pipe(filter(t => t), take(1)).subscribe(t => {
			var monaco = (window as any).monaco;
			//This allows you to set the URI for the editor
			this.editorOptions.model = monaco.editor.createModel(this.code, this.editorOptions.language, monaco.Uri.parse('file:///main.ts'));
			//This handles adding custom types and stuff
			monaco.languages.typescript.typescriptDefaults.setCompilerOptions({ noLib: true, allowNonTsExtensions: true });
			monaco.languages.typescript.typescriptDefaults.setExtraLibs([{
				content: `...`,
				filePath: 'file:///node_modules/@types/codecomplete.d.ts'
			}]);
		});
	}

	editorInit(editor: monacoEditor.editor.IStandaloneCodeEditor) {
		this.editor = editor;
	}

	ngOnInit(): void {}
}
```

**Comment by noririco** [comment]
@calico-crusade can you please show the `<ngx-monaco-editor>` part of the html ? thanks!

**Comment by nikhilnxvverma1** [comment]
> I found a work around for now, it isn't very pretty though:
> In the component with your editor, you can do this:
> 
> ```ts
> import { Component, OnInit } from '@angular/core';
> import { MonacoEditorLoaderService } from '@materia-ui/ngx-monaco-editor';
> //Using this specifically for getting typings, don't use it for actual code related things, as it isn't the same instance as window.monaco
> import * as monacoEditor from 'monaco-editor';
> import { filter, take } from 'rxjs/operators';
> 
> @Component({
> 	selector: 'cc-editor',
> 	templateUrl: './editor.component.html',
> 	styleUrls: ['./editor.component.scss']
> })
> export class EditorComponent implements OnInit {
> 	private editor: monacoEditor.editor.IStandaloneCodeEditor;
> 	editorOptions = {
> 		theme: 'vs-dark',
> 		language: 'typescript',
> 		minimap: {
> 			enabled: false
> 		},
> 		model: null
> 	};
> 	code: string = `...`;
> 
> 	constructor(
> 		private _monacoSer: MonacoEditorLoaderService
> 	) {
> 		this._monacoSer.isMonacoLoaded$.pipe(filter(t => t), take(1)).subscribe(t => {
> 			var monaco = (window as any).monaco;
> 			//This allows you to set the URI for the editor
> 			this.editorOptions.model = monaco.editor.createModel(this.code, this.editorOptions.language, monaco.Uri.parse('file:///main.ts'));
> 			//This handles adding custom types and stuff
> 			monaco.languages.typescript.typescriptDefaults.setCompilerOptions({ noLib: true, allowNonTsExtensions: true });
> 			monaco.languages.typescript.typescriptDefaults.setExtraLibs([{
> 				content: `...`,
> 				filePath: 'file:///node_modules/@types/codecomplete.d.ts'
> 			}]);
> 		});
> 	}
> 
> 	editorInit(editor: monacoEditor.editor.IStandaloneCodeEditor) {
> 		this.editor = editor;
> 	}
> 
> 	ngOnInit(): void {}
> }
> ```

Thank you. I wanted to get that stupid `noLib` thing to work and your comment above that import (monacoEditor) really made sense. Essentially, one should be doing 
`const monaco = (window as any).monaco;`

**Comment by calico-crusade** [comment]
> 
> 
> @calico-crusade can you please show the `<ngx-monaco-editor>` part of the html ? thanks!

@noririco Here is all of the HTML for that component, I apologize for the late reply. 
```HTML
<div class="full-wrapper">
	<div class="editor-wrapper">
		<ngx-monaco-editor style="height: 100%; width: 100%;"
						   [options]="editorOptions"
						   [(ngModel)]="code"
						   (init)="editorInit($event)"></ngx-monaco-editor>
	</div>
</div>
```

I ended up abandoning the project where I was using monaco due to time constraints, however, I'm happy to help if you need more information about it.

@nikhilnxvverma1 yes. The instance of monaco to use is the one defined on window. I was just using the other import to get typings to work correctly, I'm sure there are better solutions than this though. Probably defining a global type definition file that defines monaco on window so it can be referenced without the `(window as any).monaco` and while still preserving the types, that would also remove the need to reference the monaco libraries in every file that needs the editor.

**Comment by jmarc-roy** [comment]
new input uri available with 5.0.0
